### ISSUE TITLE
SqlDatabase: Failed to set owner to an ADDS group

### ISSUE THREAD

**Issue by JamesDLD** [dsccommunity/SqlServerDsc#829]
Hello,

I can't figure out why I can't assign db_owner permission on an ADDS group to a database, adding below the configuration.
The strange behavior is that it's working perfectly when assigning the db_owner permission on an ADDS user.

1. Used version
- SQL Server version : 2016
- OS : Windows 2016
- xSQLServer version : 8.1.0.0

2.  Configuration 

```powershell
Configuration CreatingDatabase 
{
    Import-DscResource -ModuleName 'xSQLServer' -ModuleVersion "8.1.0.0"

    Node $AllNodes.NodeName
    {   
        LocalConfigurationManager 
        {
            RebootNodeIfNeeded = $false
            CertificateID = $node.Thumbprint 
        }

        $Group = "CONTOSO\P_My_AD_Group"

        xSQLServerDatabase ServerDatabase
        {
            Ensure          = 'Present'
            SQLServer       = 'FRDCDWTEST01'
            SQLInstanceName = 'P12345'
            Name            = 'DBName3'
        }

        xSQLServerLogin Add_SqlServerLogin_SQLAdmin
        {
            Ensure               = 'Present'
            Name                 =  $Group
            LoginType            = 'WindowsGroup' #'WindowsUser'
            SQLServer            = 'FRDCDWTEST01'
            SQLInstanceName      = 'P12345'
            DependsOn = '[xSQLServerDatabase]ServerDatabase'
        }

        xSQLServerDatabasePermission SQLServerDatabasePermission
        {
            Ensure               = 'Present'
            Name                 = $Group
            Database             = 'DBName3'
            PermissionState      = 'Grant'
            Permissions          = 'Connect', 'Update'
            SQLServer            = 'FRDCDWTEST01'
            SQLInstanceName      = 'P12345'
            DependsOn = '[xSQLServerLogin]Add_SqlServerLogin_SQLAdmin'
        }

        xSQLServerDatabaseOwner Set_SqlDatabaseOwner_SQLAdmin
        {
            Name                 =  $Group 
            Database             = 'DBName3'
            SQLServer            = 'FRDCDWTEST01'
            SQLInstanceName      = 'P12345'
            DependsOn = '[xSQLServerDatabasePermission]SQLServerDatabasePermission'
        }
    }
}
```

2.  Error
```
PowerShell DSC resource MSFT_xSQLServerDatabaseOwner  failed to execute Set-TargetResource functionality with error message: Failed to set owner named CONTOSO\P_My_AD_Group of the database named DBName3 on FRDCDWTEST01\P12345. InnerException: Exception calling "SetOwner" with "1" argument(s): "Set owner failed 
for Database 'DBName3'. " 
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure
    + PSComputerName        : FRDCDWTEST01
 
The SendConfigurationApply function did not succeed.
    + CategoryInfo          : NotSpecified: (root/Microsoft/...gurationManager:String) [], CimException
    + FullyQualifiedErrorId : MI RESULT 1
    + PSComputerName        : FRDCDWTEST01
```

Thank you for your help
James

**Comment by johlju** [comment]
Hmm. That's strange, since you can set a user. I have to test that.

**Comment by JamesDLD** [comment]
Hello @johlju 
did you get the time to test that?
Thanks,
James

**Comment by JamesDLD** [comment]
For info, we are using script through xSQLServerScript as a workaround.

**Comment by johlju** [comment]
@JamesDLD No sorry. 😞 Feel bad about it, both this issue and other, but had to prioritize PR reviews.
Would it be possible for you to post the script you are using, if you are allowed to share it?

**Comment by JamesDLD** [comment]
No pb @johlju there is the script I used, it triggers the AD group as per a specific naming convention ==>

**Get Script**
```javascript
EXEC sp_MSForEachDB 'SELECT ''?'' AS [Database Name], su1.name AS [Database User Name], su2.name AS [Database Role]
                  FROM [?].sys.database_role_members r
                     INNER JOIN [?]..sysusers su1 ON su1.[uid] = r.member_principal_id
                     INNER JOIN [?]..sysusers su2 ON su2.[uid] = r.role_principal_id
                  WHERE su2.name IN(''db_owner'') AND su1.name NOT IN(''dbo'')'
```

**Test Script**
```javascript
IF EXISTS (select * from master.sys.server_principals where name like '%_O')
BEGIN
	PRINT 'ADD Owner user finished successfully'
END
ELSE
BEGIN
	RAISERROR ('Db Owner user doesn t exist. Errror with script 4_Set-ScomPermission-Set.sql', 16, 1)
END



DECLARE @accountname nvarchar(128)
DECLARE @Dbname nvarchar(128)
DECLARE @servername nvarchar(128)
DECLARE @command1 nvarchar(MAX)
DECLARE @command2 nvarchar(MAX)

--get the server name-----------
Select  @servername = HOST_NAME()


DECLARE db_cursor CURSOR FOR  


---define the database that must be used to add the owner user ---

SELECT name FROM master.dbo.sysdatabases where 

name not in ('master','tempdb','model','msdb','sysdba','sysmon')

-------define the environement type------------


OPEN db_cursor   

SET @command2 = ''

FETCH NEXT FROM db_cursor INTO @Dbname   


WHILE @@FETCH_STATUS = 0   
BEGIN   

 DECLARE @Username nvarchar(128)
  
------build the username string ---------------------

set @command1= 

'use ['+@Dbname+']

IF EXISTS (select name from sysusers where name like '''+@Dbname+'%_O'')
BEGIN
	PRINT ''ADD Owner user finished successfully''
END
ELSE
BEGIN
	RAISERROR (''Db Owner user doesn t exist. Errror with script 4_Set-ScomPermission-Set.sql'', 16, 1)
END
'
set @command2 = @command2 + @command1

 
FETCH NEXT FROM db_cursor INTO @Dbname   

END   

CLOSE db_cursor   

DEALLOCATE db_cursor   

EXECUTE sp_executesql @command2; 
```

**Set Script**
```javascript
--DECLARE VARIABLES

DECLARE @accountname nvarchar(128)
DECLARE @Dbname nvarchar(128)
DECLARE @EnvType nvarchar(128)
DECLARE @instancename  varchar (128)
DECLARE @instancenametemp  varchar (128)
DECLARE @instancenameENVCaract  varchar (1)
DECLARE @servername nvarchar(128)
DECLARE @command1 nvarchar(MAX)
DECLARE @command2 nvarchar(MAX)

--get the server name-----------
Select  @servername = HOST_NAME()

--get the server servername\instancename----
select @instancename = @@servername

-- intermediate operation to get the instance name -----------
select  @instancenametemp = substring(@instancename,0, charindex('\',@instancename))

---depanding on the on the kind of hostname (azure or vmware) the conventional name is different

select  @instancenameENVCaract = 
case
when @instancenametemp  like 'VM%' then   SUBSTRING (@instancenametemp, 8, 1)
when  @instancenametemp  like 'FRD%' then SUBSTRING (@instancenametemp,5, 1)

end


DECLARE db_cursor CURSOR FOR  


---define the database that must be used to add the owner user ---

SELECT name FROM master.dbo.sysdatabases where 

name not in ('master','tempdb','model','msdb','sysdba','sysmon')

-------define the environement type------------

 select  @EnvType = 
	Case 
	when  @instancenameENVCaract = 'P' then   'PROD' 
	when  @instancenameENVCaract = 'I' then   'INT' 
	when  @instancenameENVCaract = 'R' then   'REC' 
	when  @instancenameENVCaract = 'D' then   'DEV' 
	END

OPEN db_cursor   

SET @command2 = ''

FETCH NEXT FROM db_cursor INTO @Dbname   


WHILE @@FETCH_STATUS = 0   
BEGIN   

 DECLARE @Username nvarchar(128)
  
------build the username string ---------------------

set @Username =  'CRITEOIS\P_'+@EnvType+'-SQL-'+@servername+'_'+@Dbname+'_O'
	 

SET @command1 = 'USE ['+@Dbname+'];
	

IF NOT EXISTS (SELECT * FROM sys.database_principals WHERE name = N'''+@Username+''')
 
CREATE USER ['+@Username+'] FOR LOGIN  ['+@Username+'] WITH DEFAULT_SCHEMA=[dbo]


ALTER ROLE [db_owner] ADD MEMBER ['+@Username+']


'
set @command2 = @command2 + @command1

 
FETCH NEXT FROM db_cursor INTO @Dbname   

END   

CLOSE db_cursor   

DEALLOCATE db_cursor   

EXECUTE sp_executesql @command2;
 
```

**Comment by johlju** [comment]
I see in the script that you are adding the user to the database role 'db_owner' which is not the same as setting the database owner. To add a user to a database role, you can use the resource 'SqlDatabaseRole'.

But the reason for not being able to set the database owner using SqlDatabaseOwner is probably because of that you add the user to the database using SqlDatabasePermission.

>login cannot become the owner of the current database if it already has access to the database through an existing user security account within the database. To avoid this, drop the user within the current database first. 
>https://docs.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-changedbowner-transact-sql

*The resource names differ from your configuration since the resources changed names in version 10.0.0.0.*

**Comment by JamesDLD** [comment]
I think I tried that but I can retry it, the strange thing is that it was working for a user but not for a group.
Did you try it just to see if I one of my code env. was wrong?
Thanks

**Comment by johlju** [comment]
Ah! I think there cannot be a group that is database owner, since the database owner maps to the 'dbo' user in the database. Trying to find information on this, since I'm not 100% certain.

**Comment by johlju** [comment]
The error message in SSMS says the correct error. 

![image](https://user-images.githubusercontent.com/7189721/34682952-e190677a-f4a0-11e7-891a-b17f55a1110b.png)

**Comment by johlju** [comment]
Thought the code maybe don't write out the `InnerExcepton`, but it did. 

https://github.com/PowerShell/SqlServerDsc/blob/b699be3aa1bfef2c9707ef8d3a54e8d2c0c67994/DSCResources/MSFT_SqlDatabaseOwner/MSFT_SqlDatabaseOwner.psm1#L149-L153

Maybe the error message we wanted to see is further in (nested exception). Maybe it's possible to show it.
But at least we should update the documentation for this resource so it mentions that only a user is supported as the database owner.

Labeling this as enhancement and help wanted so someone can run with this issue.

**Comment by johlju** [comment]
I'm closing this issue as the resource SqlDatabase has been refactored to support the Owner property. It will be released in the v14.0.0. If there is still an issue with that resource we can reopen this issue.
